[SofaKernel] FIX in TetrahedronFEMForceField & TetrahedronSetTopologyAlgorithm

damienmarchal opened the pull request:
Hi, here is the content:

1) [SofaKernel] FIX segfault in TetrahedronSetGeometryAlgorithms
The segfault happens when the component is located in a node without
a topology and a mechanical object. Then the 'draw' function is crashing.

2)  [SofaKernel] ADD a link to manually set the topology in TetrahedronFEMForceField
The component was searching the topology in the context which was:
- implicit
- constraining the organization of components the scene graph.

To solve this I added a SingleLink that allows to specify the topology.
If the link is not set then the existing behavior is used.

______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

damienmarchal committed:
[SofaKernel] FIX segfault in TetrahedronSetGeometryAlgorithms

The segfault happens when the component is located in a node without
a topology and a mechanical object. Then the 'draw' function is crashing.

damienmarchal committed:
[SofaKernel] ADD a link to manually set the topology in TetrahedronFEMForceField

The component was searching the topology in the context which was:
- implicit
- constraining the organization of components the scene graph.

To solve this I added a SingleLink that allows to specify the topology.
If the link is not set then the existing behavior is used.

epernod commented:
there are new tests failure:
```

FastTriangularBendingSprings.scn:0, Dynamic Parser, Priority: Normal
--
[EdgeSetGeometryAlgorithms(edgeSetGeometryAlgorithms1)] Unable to get a valid topology from the context

```

damienmarchal committed:
[SofaConstraint] FIX invalid #ifdef in BilateralInteractionConstraint.

The checked ifdef is not valid and not following the SOFA way of doing
things. So I fixed it.

damienmarchal committed:
[SofaKernel] FIX BarycentricMapping crashing SOFA when there is no topology.

damienmarchal committed:
Fix the broken scene.

damienmarchal commented:
Thank for the reviewed. 
I checked the example file and my thinking is that this example is wrong and my PR is legitimate. 
1) if you change in the example to turn visualization of the edge set it will crash. 
2) The edgeset is in the root node and so the existing code to retrieve the context in the init() is not working as we expect.
3) The code in the PR is right at detecting and reporting an error when this case is detected. 

So i moved the EdgeSet into the same node as the topology. So the error does not pop up.

epernod commented:
yes, a EdgeSetGeometryAlgorithms should not be inside a node without mechanicalObject nor a topology container (I know a lot of check are still missing :( ). 
I'm even surprise, it works with this: 
```
                <MeshTopology src="@loader" />
                <EdgeSetGeometryAlgorithms />
```

epernod committed:
[scenes] Fix broken scene. EdgeSetGeometry should be in the node with mechanicalObject
